homeassistant/components/stream/hls.py

```diff
@@ -0,0 +1,128 @@
+    name = 'api:stream:hls:playlist'
+    cors_allowed = True
+
+    async def handle(self, request, stream, sequence):
+        """Return m3u8 playlist."""
+        renderer = M3U8Renderer(stream)
+        track = stream.add_provider(HlsStreamOutput())
+        stream.start()
+        # Wait for a segment to be ready
+        if not track.segments:
+            await track.recv()
+        headers = {
+            'Content-Type': 'application/vnd.apple.mpegurl'
+        }
+        return web.Response(body=renderer.render(
+            track, datetime.datetime.utcnow()
+            ).encode("utf-8"), headers=headers)
+
+
+class HlsSegmentView(StreamView):
+    """Stream view to serve a MPEG2TS segment."""
+
+    url = r'/api/hls/{token:[a-f0-9]+}/segment/{sequence:\d+}.ts'
+    name = 'api:stream:hls:segment'
+    cors_allowed = True
+
+    async def handle(self, request, stream, sequence):
+        """Return mpegts segment."""
+        track = stream.add_provider(HlsStreamOutput())
+        segment = track.get_segment(int(sequence))
+        if not segment:
+            return web.HTTPNotFound()
+        headers = {
+            'Content-Type': 'video/mp2t'
+        }
+        return web.Response(body=segment.segment.getvalue(), headers=headers)
+
+
+class M3U8Renderer:
+    """M3U8 Render Helper."""
+
+    def __init__(self, stream, version=3):
+        """Initialize renderer."""
+        self.stream = stream
+        self.version = version
+
+    def render_preamble(self, track):
+        """Render preamble."""
+        return [
+            "#EXT-X-VERSION:{}".format(self.version),
+            "#EXT-X-TARGETDURATION:{}".format(track.target_duration),
+        ]
+
+    def render_segment(self, segment):
+        """Render segment."""
+        return [
+            "#EXTINF:{:.04},".format(float(segment.duration)),
+            "/api/hls/{}/segment/{}.ts".format(
+                self.stream.access_token,
+                segment.sequence),
+        ]
+
+    def render_playlist(self, track, start_time):
+        """Render playlist."""
+        segments = track.segments
+
+        if not segments:
+            return []
+
+        playlist = ["#EXT-X-MEDIA-SEQUENCE:{}".format(segments[0])]
+
+        for sequence in segments:
+            segment = track.get_segment(sequence)
+            playlist.extend(self.render_segment(segment))
+
+        return playlist
+
+    def render(self, track, start_time):
+        """Render M3U8 file."""
+        lines = (
+            ["#EXTM3U"] +
+            self.render_preamble(track) +
+            self.render_playlist(track, start_time)
+        )
+        return "\n".join(lines) + "\n"
+
+
+class HlsStreamOutput(StreamOutput):
+    """Represents HLS Output formats."""
+
+    @property
+    def format(self) -> str:
+        """Return container format."""
+        return 'mpegts'
+
+    @property
+    def audio_codec(self) -> str:
+        """Return desired audio codec."""
+        return 'aac'
```

Comment:
So this is beyond MVP, more a question:

Would it be possible that a stream output can define multiple supported codecs?

Reply:
Yes.  I was actually thinking of doing that if I build a WebRTC provider.  That supports VP8 and H264 depending on browser, as well as Opus and G.711.

A WebRTC provider would enable streaming to Alexa devices as well as real-time streaming in supported browsers instead of the buffered delays of HLS, but was determined to be out of scope for MVP.

The implementation would probably be to define the codecs in this method as a list, check if the source stream is already in that format.
- If not, transcode to the first format in the provided list.
- Otherwise, copy the source stream without transcoding.